ste/JobsAdmin.go

```diff
@@ -455,6 +455,7 @@ func (ja *jobsAdmin) transferProcessor(workerID int) {
 			if jptm.ShouldLog(pipeline.LogInfo) {
 				jptm.Log(pipeline.LogInfo, fmt.Sprintf(" is not picked up worked %d because transfer was cancelled", workerID))
 			}
+			jptm.SetStatus(common.ETransferStatus.Failed())
```

Comment:
1. I see that we do not specially handle cancelled as different status from Failed(). These changes were necessary because, in the existing workflow we leave the status to be Started(), which interferes with the bookkeeping and technically wrong. Other reason, with the existing code, there can be cancelled jptms with both Failed() and Started() status, depending on where they failed. Marking the Started() jptms Failed() would make it uniform. This might be a good time to introduce Cancelled() status - it can inform us the reason for failure of this file (similar to other status like BlobTierFailure()), but I'm not sure it might be useful.
2. Now assuming Started() is not a terminal/Static state for jptm, I expected the the transferStatus terminal state (intutively) before we ReportTransferDone(). And the places where I've set the status to Failed(), violated this condition. There were two places this violation could have been fixed: 
	a. Before calling ReportTransferDone(), set the status appropriately
	b. At the places where the status actually changes - like the Started() when scheduling, Completed() on success in Epilogue, Failed() on failure, and *Failed() when the jptm is cancelled*
3. I totally ignored the point (b) above until your comments(KMN). The new places where we've set the status dees not impact Cancel() and WasCancelled(). But all the other forms of isDead() invoked after SetJobStatus() line would behave differently (Though they still return the same result - Need to verify this again). 

I'm inclined towards moving jptm.TransferStatus to a terminal status with this change. I'm looking at 2a and 2b bit more. I think both don't impact Is*, though 2a appears to.

Reply:
Update:
I looked at the code to understand a bit more about how we cancel a job and what might be a good place to set the status to Failed().

How do we cancel the job? This is required because our jptm.isDead() checks the context to decide.
All the jptm are given a new context (jptm.Ctx) whose parent is JobMgr's context(jm.Ctx). The jptm.Ctx is released either by invoking jptm.Cancel() or jm.Cancel(), whichever happens first. When transfer of a file is complete(failure or success), the jptm.Ctx is released by invoking jptm.Cancel. When we cancel a job, jptm.Ctx is released because parent ctx is released.

Where can we set the transfer Status to be Failed()
1. Reactive: Before calling ReportTransferDone(), if we find the status is not terminal, we set it appropriately. This is what I'd implemented before this discussion. In terms of code, we'd have to change the status to Failed() in commonSenderCompletion(), commonDownloaderCompletion(). Not a fan of this method but appears clean, because we set the Success() state here.

2. reactive: When we release jptm.Ctx, because the parent context - jm.Ctx is released. This would require us to loop through all the jptms while we are cancelling jobs, should not be done. 

3. First Detection: Set the status to Failed() when the jptm can detect its context is released.
Now, when can we detect the jptm.Ctx is released:
a. Before we schedule chunks - remoteToLocal(), anyToRemote()
b. After we schedule chunks - epilogueWithCleanupDownload(), epilogueWithCleanupSendToRemote()

In any case, we'd have to set the status in transferProcessor() because of a possible race. Also, jptm.Is* methods behave similarly in all the cases. As soon as the context is released, jptm is considered dead.